openwisp_firmware_upgrader/tasks.py

```diff
@@ -83,3 +83,27 @@ def create_all_device_firmwares(self, firmware_image_id):
     queryset = Device.objects.filter(os=fw_image.build.os)
     for device in queryset.iterator():
         DeviceFirmware.create_for_device(device, fw_image)
+
+
+@shared_task(base=OpenwispCeleryTask)
+def delete_firmware_files(files_to_delete):
+    FirmwareImage = load_model("FirmwareImage")
+    storage = FirmwareImage.file.field.storage
+
+    for file_path in files_to_delete:
+        try:
+            storage.delete(file_path)
+            dir_path = "/".join(file_path.split("/")[:-1])
+            if dir_path:
+                try:
+                    dirs, files = storage.listdir(dir_path)
+                    if not dirs and not files:
+                        storage.delete(dir_path)
+                    else:
+                        logger.debug(
+                            f"Directory {dir_path} is not empty, skipping deletion"
```

Comment:
Let’s use `%-style` formatting with `logger.*()` since the Python logging library is optimized for it - it defers string interpolation until it’s actually needed (based on the log level), which is more efficient. 
https://stackoverflow.com/questions/54367975/python-3-7-logging-f-strings-vs

Reply:
ex:
```py
logger.debug("Directory %s is not empty, skipping deletion", dir_path)
```